### PR TITLE
backport: guix related backports

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
This PR improve overall support of guix:
 - bump some packages versions
 - improve guix scripts
 - introduce Mac OS X builds and many other things.

Furthermore Guix is considered as a replacement for gitian for binary builds.

Prior work: https://github.com/dashpay/dash/pull/5194
Related issue: https://github.com/dashpay/dash-issues/issues/47

## What was done?
There is a bunch of backports that is related to guix. These backports improve overall support of guix

- bitcoin/bitcoin#18958
- bitcoin/bitcoin#20318
- bitcoin/bitcoin#20619
- bitcoin/bitcoin#20937
- bitcoin/bitcoin#17920
- bitcoin/bitcoin#20629
- bitcoin/bitcoin#21116
- bitcoin/bitcoin#21163
- bitcoin/bitcoin#21087
- bitcoin/bitcoin#21088
- partial bitcoin/bitcoin#21089
- bitcoin/bitcoin#21078
- bitcoin/bitcoin#21271
- bitcoin/bitcoin#21272
- bitcoin/bitcoin#21321
- bitcoin/bitcoin#21323
- bitcoin/bitcoin#21298
- partial bitcoin/bitcoin#21325
- bitcoin/bitcoin#21337
- bitcoin/bitcoin#21375
- bitcoin/bitcoin#21611
- partial bitcoin/bitcoin#21626
- bitcoin/bitcoin#21304
- bitcoin/bitcoin#21661
- bitcoin/bitcoin#21672
- bitcoin/bitcoin#21655
- bitcoin/bitcoin#19817
- bitcoin/bitcoin#21799
- bitcoin/bitcoin#21462
- bitcoin/bitcoin#21239
- bitcoin/bitcoin#21654
- bitcoin/bitcoin#22075
- bitcoin/bitcoin#22190
- bitcoin/bitcoin#22182
- bitcoin/bitcoin#22365
- partial bitcoin/bitcoin#22405
- bitcoin/bitcoin#22410
- bitcoin/bitcoin#22465
- bitcoin/bitcoin#22533
- bitcoin/bitcoin#22643
- bitcoin/bitcoin#22847
- [TO BE REVERTED LATER] bitcoin/bitcoin#24215
- bitcoin/bitcoin#18395

## What is NOT done yet?
 - there is still no dashification, it would be done in further PR after backport bitcoin/bitcoin#21711
 - currently builds for PowerPC fails, most likely failure is related to outdated `freetype` package. To-be-fixed in further backports
 - backport bitcoin/bitcoin#24215 should be reverted because the proper way to do it to bump `guix` to newer version. It would be done in futher PR, not trivial to be done ATM.
 - backport bitcoin/bitcoin#22405 is partial, because we do not pass flag "--enable-glibc-back-compat" for all targets how it is done for bitcoin. We use it for gitian, for `native_qt5` and `native_sqlite`, but doesn't have for `native_tsan`, `arm`, `native_cxx20`. Instead, let's do better bitcoin/bitcoin#22930 and after that we can finish this backport bitcoin#22405 by adding missing changes:
```diff
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -943,7 +943,7 @@ if TARGET_WINDOWS
        $(AM_V_at) OBJDUMP=$(OBJDUMP) $(PYTHON) $(top_srcdir)/contrib/devtools/symbol-check.py $(bin_PROGRAMS)
 endif
 
-if GLIBC_BACK_COMPAT
+if TARGET_LINUX
        @echo "Checking glibc back compat..."
        $(AM_V_at) CPPFILT=$(CPPFILT) $(PYTHON) $(top_srcdir)/contrib/devtools/symbol-check.py $(bin_PROGRAMS)
 endif
```

## How Has This Been Tested?
 - Firstly, you need to get SDK path for Mac OS build. Official recommended way is download it from Apple's website using Apple ID
 - set up environment variable `export SDK_PATH=....`. Be aware, that it works only for full path (not relative). Otherwise you will see this error similar to:
```
clang-10: warning: no such sysroot directory: 'SDK_PATH/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers' [-Wmissing-sysroot]
In file included from libs/thread/src/pthread/thread.cpp:9:
In file included from ./boost/thread/detail/config.hpp:11:
In file included from ./boost/config.hpp:44:
./boost/config/detail/select_stdlib_config.hpp:26:14: fatal error: 'cstddef' file not found
#    include <cstddef>
```
 - run `contrib/guix/guix-build`. You need ~30Gb of free space to get built all platform.
NOTE: be careful with `./contrib/guix/guix-clean` because it removes all files without an entry in the index.

# Guix hashes
```
$ find guix-build-"$(git  rev-parse --short=12 c09129bfff55ee82861998d985f2bc49cc498a6c)"/output -type f | grep -v SHA256 | xargs  sha256sum  | sort -k 2
ef22745c6f4674c314af2e8d47c3b2078025041be7363eaf5a80d2ebe4183cef  guix-build-c09129bfff55/output/aarch64-linux-gnu/bitcoin-c09129bfff55-aarch64-linux-gnu-debug.tar.gz
6901f13b487a0385d3c441f243830d3ee04a070baced5f5a041eb6f6760dafa5  guix-build-c09129bfff55/output/aarch64-linux-gnu/bitcoin-c09129bfff55-aarch64-linux-gnu.tar.gz
a5c19915a39b096b681efd14e649fddc930b9e676a9ebe0d60d04f3eaa368ff7  guix-build-c09129bfff55/output/arm-linux-gnueabihf/bitcoin-c09129bfff55-arm-linux-gnueabihf-debug.tar.gz
0eed27a58acbc0d11776e6950c3f716807aade0d05542b79ae342a23389541ea  guix-build-c09129bfff55/output/arm-linux-gnueabihf/bitcoin-c09129bfff55-arm-linux-gnueabihf.tar.gz
1f86baf527062c7f47ec567fd0ec55f9e104f7ed99c4273fec463e7a086376f1  guix-build-c09129bfff55/output/dist-archive/bitcoin-c09129bfff55.tar.gz
515c071f63c6bd0e9ca1baeea121ff000e47303c507e15a6a5e9db1ae68823bf  guix-build-c09129bfff55/output/riscv64-linux-gnu/bitcoin-c09129bfff55-riscv64-linux-gnu-debug.tar.gz
1a82c21c72126f0dde54cb772fb6f2d1a877982c70c194608fe02ad4a1696a56  guix-build-c09129bfff55/output/riscv64-linux-gnu/bitcoin-c09129bfff55-riscv64-linux-gnu.tar.gz
425808ce4e1ffe658ad3a51055d35aebe33bbeba02003a4d9a6301b7028a3629  guix-build-c09129bfff55/output/x86_64-apple-darwin18/bitcoin-c09129bfff55-osx64.tar.gz
174619e05109b1c84d18fd4ae34cfc330cd51e4ce8616e919f636abe512c4c52  guix-build-c09129bfff55/output/x86_64-apple-darwin18/bitcoin-c09129bfff55-osx-unsigned.dmg
378fb01c9788217e545cc1de0c376bcd5465d2712ad760c885b924d666a19c0f  guix-build-c09129bfff55/output/x86_64-apple-darwin18/bitcoin-c09129bfff55-osx-unsigned.tar.gz
6e6913fb45e8ba2c17af80746666d4d770f5d06e7eea9bac2b54803115393186  guix-build-c09129bfff55/output/x86_64-linux-gnu/bitcoin-c09129bfff55-x86_64-linux-gnu-debug.tar.gz
a9f335769802fb1e57dc54f74f28a4926f05477075b54cc199e5c13c7d34f183  guix-build-c09129bfff55/output/x86_64-linux-gnu/bitcoin-c09129bfff55-x86_64-linux-gnu.tar.gz
6f10b508277d30d358548c591f83c6050709faa210715c241c02a64b6d853f49  guix-build-c09129bfff55/output/x86_64-w64-mingw32/bitcoin-c09129bfff55-win64-debug.zip
70c1e6d3278c9ef74c05b1f8ce164c3e5d4c90bc1283edd04981f453861f4948  guix-build-c09129bfff55/output/x86_64-w64-mingw32/bitcoin-c09129bfff55-win64-setup-unsigned.exe
562accd3e4c8e37731aa1077a2d7dfe7c9ac6718c1f77864092995cd21970401  guix-build-c09129bfff55/output/x86_64-w64-mingw32/bitcoin-c09129bfff55-win64.zip
75b1c423dff080846176b26ab423173bb85ecfae291d8a932b9adf67ca38f4be  guix-build-c09129bfff55/output/x86_64-w64-mingw32/bitcoin-c09129bfff55-win-unsigned.tar.gz

```
# Gitian hashes
Don't forget to add [Xcode-12.1](https://bitcoincore.org/depends-sources/sdks/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers.tar.gz) to `gitian-builder/inputs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers.tar.gz`
```
$ ./dash/contrib/gitian-build.py -b -n -j 20 -m 12000 knst 19-guix -u git@github.com:knst/dash
1e1d9be26845a230628f9396d257ca8fb29cfab4a70c4926ea79b0e495d36216  dashcore-19-guix-aarch64-linux-gnu-debug.tar.gz
183692c239a8ff5ae3c80d5f33a9350cb5c8a2289416b03b6a78d3b3ac3ed6ef  dashcore-19-guix-aarch64-linux-gnu.tar.gz
cbd52e8dacd36575127495674295d6ba4e05029661f4249b5f99edad465d9168  dashcore-19-guix-riscv64-linux-gnu-debug.tar.gz
ce2d4bcfa293fcb63ece79b394c0fc11869b2dda65d7edeeb4c036091d762fe1  dashcore-19-guix-riscv64-linux-gnu.tar.gz
bce1f2bc4e0f9da2a04460e55257751ef66c688e1b9c272eded03a112497881e  dashcore-19-guix-x86_64-linux-gnu-debug.tar.gz
7c9dc9ffc70adae26868d7e92ae32d8640ab289b953bfa35eb34dd867587618d  dashcore-19-guix-x86_64-linux-gnu.tar.gz
96c27de90527975414f9f11fa789d59d651c0406eb48095afbae9b709b7553b2  src/dashcore-19-guix.tar.gz
b43b950a86cae9b0dd7844b0fd027ddd3e2fd087513e8d54f3b92576a78eec1a  dash-linux-19-res.yml

f83bd235df11cb74c5b0f6dc738f678690b83c3be605fdf3dfbde4704fc46887  dashcore-19-guix-win-unsigned.tar.gz
148caee2c0c727d702e213ef79ad7a4ea0daeca15c042caec6df5c8d55049ed1  dashcore-19-guix-win64-debug.zip
aed6889fb1edf016dde2878ed9a3c3ce3f12ea360ea7d64f86141b8c373d394b  dashcore-19-guix-win64-setup-unsigned.exe
1c06e43724dd28e81279b9261bb890db0f042f475e1dcf6746758a0b43f227e0  dashcore-19-guix-win64.zip
96c27de90527975414f9f11fa789d59d651c0406eb48095afbae9b709b7553b2  src/dashcore-19-guix.tar.gz
6e73f69d6182ec002530c804ff706ebc599ce84bbce151d12c55462631045a82  dash-win-19-res.yml

bb4c960f322208a372fd5fcd7b9254e85b5e4eac156c851ca15d838ad7cc1ebf  dashcore-19-guix-osx-unsigned.dmg
392964c3c77b5bc10a9a9b6d53f19c4a2a2766962fde211f79abd1373c21beda  dashcore-19-guix-osx-unsigned.tar.gz
08dc7e51731df9acf3ff5f7606ed3fac76e0984497be836caa03fa6ee0ed5f1f  dashcore-19-guix-osx64-debug.tar.gz
d9194881f4484f82982df0de527f6f57b0671fbabf2c56f74e96435d905ff33c  dashcore-19-guix-osx64.tar.gz
96c27de90527975414f9f11fa789d59d651c0406eb48095afbae9b709b7553b2  src/dashcore-19-guix.tar.gz
125c824946d2bab67f1e6397aab68d37e17fce170e1cc34fb5cf58d40428e9dd  dash-osx-19-res.yml
```

## Breaking Changes
Should not be breaking changes at the moment. Btw, there's bump of XCode sdk version.

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation - see a chapter "TODO"
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by knst** [comment]
`contrib/shell/realpath.bash`
> # Based on realpath.sh written by Michael Kropat
> # Found at: https://github.com/mkropat/sh-realpath/blob/65512368b8155b176b67122aa395ac580d9acc5b/realpath.sh

@thephez 
Should we add these credits somewhere else in docs/licenses/etc? Seems as it is licensed by MIT license in the original repo.

**Comment by thephez** [comment]
> `contrib/shell/realpath.bash`
> 
> > Based on realpath.sh written by Michael Kropat
> > Found at: https://github.com/mkropat/sh-realpath/blob/65512368b8155b176b67122aa395ac580d9acc5b/realpath.sh
> 
> @thephez Should we add these credits somewhere else in docs/licenses/etc? Seems as it is licensed by MIT license in the original repo.

Did this come via a backport? If so, we should probably just handle it however bitcoin did.

**Comment by PastaPastaPasta** [comment]
my results 
```
pasta@pasta1 /mnt/volume_tor1_01/dash
 % find guix-build-"$(git  rev-parse --short=12 c09129bfff55ee82861998d985f2bc49cc498a6c)"/output -type f | grep -v SHA256 | xargs  sha256sum  | sort -k 2

ef22745c6f4674c314af2e8d47c3b2078025041be7363eaf5a80d2ebe4183cef  guix-build-c09129bfff55/output/aarch64-linux-gnu/bitcoin-c09129bfff55-aarch64-linux-gnu-debug.tar.gz
6901f13b487a0385d3c441f243830d3ee04a070baced5f5a041eb6f6760dafa5  guix-build-c09129bfff55/output/aarch64-linux-gnu/bitcoin-c09129bfff55-aarch64-linux-gnu.tar.gz
a5c19915a39b096b681efd14e649fddc930b9e676a9ebe0d60d04f3eaa368ff7  guix-build-c09129bfff55/output/arm-linux-gnueabihf/bitcoin-c09129bfff55-arm-linux-gnueabihf-debug.tar.gz
0eed27a58acbc0d11776e6950c3f716807aade0d05542b79ae342a23389541ea  guix-build-c09129bfff55/output/arm-linux-gnueabihf/bitcoin-c09129bfff55-arm-linux-gnueabihf.tar.gz
1f86baf527062c7f47ec567fd0ec55f9e104f7ed99c4273fec463e7a086376f1  guix-build-c09129bfff55/output/dist-archive/bitcoin-c09129bfff55.tar.gz
515c071f63c6bd0e9ca1baeea121ff000e47303c507e15a6a5e9db1ae68823bf  guix-build-c09129bfff55/output/riscv64-linux-gnu/bitcoin-c09129bfff55-riscv64-linux-gnu-debug.tar.gz
1a82c21c72126f0dde54cb772fb6f2d1a877982c70c194608fe02ad4a1696a56  guix-build-c09129bfff55/output/riscv64-linux-gnu/bitcoin-c09129bfff55-riscv64-linux-gnu.tar.gz
174619e05109b1c84d18fd4ae34cfc330cd51e4ce8616e919f636abe512c4c52  guix-build-c09129bfff55/output/x86_64-apple-darwin18/bitcoin-c09129bfff55-osx-unsigned.dmg
378fb01c9788217e545cc1de0c376bcd5465d2712ad760c885b924d666a19c0f  guix-build-c09129bfff55/output/x86_64-apple-darwin18/bitcoin-c09129bfff55-osx-unsigned.tar.gz
425808ce4e1ffe658ad3a51055d35aebe33bbeba02003a4d9a6301b7028a3629  guix-build-c09129bfff55/output/x86_64-apple-darwin18/bitcoin-c09129bfff55-osx64.tar.gz
6e6913fb45e8ba2c17af80746666d4d770f5d06e7eea9bac2b54803115393186  guix-build-c09129bfff55/output/x86_64-linux-gnu/bitcoin-c09129bfff55-x86_64-linux-gnu-debug.tar.gz
a9f335769802fb1e57dc54f74f28a4926f05477075b54cc199e5c13c7d34f183  guix-build-c09129bfff55/output/x86_64-linux-gnu/bitcoin-c09129bfff55-x86_64-linux-gnu.tar.gz
75b1c423dff080846176b26ab423173bb85ecfae291d8a932b9adf67ca38f4be  guix-build-c09129bfff55/output/x86_64-w64-mingw32/bitcoin-c09129bfff55-win-unsigned.tar.gz
6f10b508277d30d358548c591f83c6050709faa210715c241c02a64b6d853f49  guix-build-c09129bfff55/output/x86_64-w64-mingw32/bitcoin-c09129bfff55-win64-debug.zip
70c1e6d3278c9ef74c05b1f8ce164c3e5d4c90bc1283edd04981f453861f4948  guix-build-c09129bfff55/output/x86_64-w64-mingw32/bitcoin-c09129bfff55-win64-setup-unsigned.exe
562accd3e4c8e37731aa1077a2d7dfe7c9ac6718c1f77864092995cd21970401  guix-build-c09129bfff55/output/x86_64-w64-mingw32/bitcoin-c09129bfff55-win64.zip
```

**Comment by knst** [comment]
> my results
hashes matched, but somehow `|sort -k2` sorted hashes in different order

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

File: contrib/gitian-build.py [hunk 1]
```diff
@@ -217,6 +217,11 @@ def main():
     args.windows = 'w' in args.os
     args.macos = 'm' in args.os
 
+    # Disable for MacOS if no SDK found
+    if args.macos and not os.path.isfile('gitian-builder/inputs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers.tar.gz'):
+        print('Cannot build for MacOS, SDK does not exist. Will build for other OSes')
+        args.macos = False
+
```
Comment (UdjinM6) [inline contrib/gitian-build.py:224]: 19817: should drop this. we do not skip macos build if no sdk is found, we download sdk instead and continue (see line 77).

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Review by UdjinM6 — APPROVED** [review]
ACK

**Pull request merged**